codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -57,22 +58,35 @@ public DefaultHttp2RemoteFlowController(Http2Connection connection) {
         connection.addListener(new Http2ConnectionAdapter() {
             @Override
             public void onStreamAdded(Http2Stream stream) {
-                // Just add a new flow state to the stream.
-                stream.remoteFlowState(new DefaultFlowState(stream, 0));
+                // If the stream state is not open then the stream is not yet eligible for flow controlled frames and
+                // only requires the ReducedFlowState. Otherwise the full amount of memory is required.
+                stream.remoteFlowState(stream.state() == IDLE ?
+                        new ReducedFlowState(stream) :
+                        new DefaultFlowState(stream, initialWindowSize));
             }
 
             @Override
             public void onStreamActive(Http2Stream stream) {
-                // Need to be sure the stream's initial window is adjusted for SETTINGS
-                // frames which may have been exchanged while it was in IDLE
-                state(stream).window(initialWindowSize);
+                // If the object was previously created, but later activated then we have to ensure
+                // the full state is allocated and the proper initialWindowSize is used.
+                if (stream.remoteFlowState().getClass() != DefaultFlowState.class) {
+                    stream.remoteFlowState(new DefaultFlowState(state(stream), initialWindowSize));
+                }
```

Comment:
I think you need an `else` where you make sure the window is set to the current value of `initialWindowSize`.